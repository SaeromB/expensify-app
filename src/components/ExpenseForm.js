import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';
// import 'react-dates/initialize';
// import 'react-dates/lib/css/_datepicker.css';

// const date = new Date();

// const now = moment();
// console.log(now.format('MMM Do, YYYY'));

export default class ExpenseForm extends React.Component {
  constructor(props){
    super(props);

    this.state = {
      // if props.expese exist then show props.expense.description if not show empty string.
      description: props.expense ? props.expense.description : '',
      note: props.expense ? props.expense.note : '',
      // we need the amount with a decimal and convert to String
      amount: props.expense ? (props.expense.amount / 100).toString() : '',
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: ''
    };
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState (() => ({description}));
  }

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState (() => ({note}))
    // e.parsist()
    // this.setState({note: e.target.value})
  };

  // will return a date when it is clicked
  onDateChange = (createdAt) => {
    if(createdAt){
      this.setState(() => ({createdAt}))
    }
  };
  onFocusChange = ({focused}) => {
    this.setState(()=> ({calendarFocused : focused}))
  }

  onAmountChange = (e) => {
    const amount = e.target.value;
    // match(regex expressions)
    // if there is no amount or the amount is a match
    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
      this.setState(() => ({amount}))
    }
  }
  
  onSubmit = (e) => {
    e.preventDefault();
    
    if(!this.state.description || !this.state.amount){
      this.setState(() => ({error: 'Please provide description and amount'})) 
    } else {
      this.setState(()=>({error: ''}))
      // this props is passed down to AddExpensePage
      // Passed the data out of Expense form
      // props is passed in to the parent?
      // this makes use the form to be reused at AddExpensePage and EditExpensePage
      this.props.onSubmit({
        description: this.state.description,
        // amount is a string working at base 10
        amount: parseFloat(this.state.amount, 10 ) * 100,
        // this is a momnet object 
        createdAt: this.state.createdAt.valueOf(),
        note: this.state.note
      }) 
      // console.log('Submitted');
    }
  }
  
  render(){
    return( 
      <div>
        {this.state.error &&  <p>{this.state.error}</p>}
        <form onSubmit={this.onSubmit}> 
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value = { this.state.description} onChange={this.onDescriptionChange} />
          <input
            type="text"
            placeholder="Amount"
            value = {this.state.amount} onChange={this.onAmountChange} />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calendarFocused}
            onFocusChange={this.onFocusChange}
            numberOfMonths={1}
            // this will make everysingle day avalible 
            isOutsideRange={()=> false }
            />
          <textarea
            placeholder=" Add a note for your expenses (optional)"
            value = {this.state.note} onChange={this.onNoteChange} >
          </textarea>
          <button>Add Expense</button>
        </form>
      </div>
    )
  }
}