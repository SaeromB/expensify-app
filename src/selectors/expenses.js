// Get visible expenses
const getVisibleExpenses = (expenses, { text, sortBy, startDate, endDate }) => {
  return expenses.filter((expense) => {
    // if the typeof startDate is not number this will be filtered if not expense.createdAt(ex:1) will be checked if it is bigger then startDate
    const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate;
    const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
    // if all of these are true and the item will be in the array if the result is false it will not show 
    return startDateMatch && endDateMatch && textMatch;
  }).sort((a, b) => {
    if(sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if(sortBy === 'amount'){
      return a.amount < b.amount ? 1 : -1
    }
  });
};

export default getVisibleExpenses;