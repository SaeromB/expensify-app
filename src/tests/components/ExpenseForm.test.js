import React from 'react';
import { shallow } from 'enzyme';
import ExpenseForm from '../../components/ExpenseForm';
import expenses from '../fixtures/expenses';
import '../setupTest';

test('Should render ExpenseForm correctly', () => {
  const wrapper = shallow(<ExpenseForm />);
  expect(wrapper).toMatchSnapshot();
});

test('Should render ExpenseForm correctly with expense data', () => {
  const wrapper = shallow(<ExpenseForm expense={expenses[1]}/>);
  expect(wrapper).toMatchSnapshot();
})

test('Should render error for invalid form submission', () => {
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {}
  });
  // getting the length of the error field.
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  // after the error state changes it gets rendered 
  expect(wrapper).toMatchSnapshot()
});
test('Should set description on input change', () => {
  const value = 'New description'
  const wrapper = shallow(<ExpenseForm />)
  wrapper.find('input').at(0).simulate('change', {
    target: { value }
  });
  expect(wrapper.state('description')).toBe(value);
});

test('should set note on textarea change', () => {
  const value = 'New Note'
  const wrapper = shallow(<ExpenseForm />)
  wrapper.find('textarea').simulate('change', {
    target: { value }
  });
  expect(wrapper.state('note')).toBe(value)
})

test('Should set amount if valid input', () => {
  const value = '23.50';
  const wrapper = shallow(<ExpenseForm />)
  wrapper.find('input').at(1).simulate('change', {
    target: { value }
  });
  // Check if the amount equals with the value we set
  expect(wrapper.state('amount')).toBe(value);
});

test('Should not set amount if valid input', () => {
  const value = '12.122';
  const wrapper = shallow(<ExpenseForm />)
  wrapper.find('input').at(1).simulate('change', {
    target: { value }
  });
  expect(wrapper.state('amount')).toBe('');
});
