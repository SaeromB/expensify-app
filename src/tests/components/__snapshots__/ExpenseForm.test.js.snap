// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ExpenseForm correctly 1`] = `ShallowWrapper {}`;

exports[`Should render ExpenseForm correctly with expense data 1`] = `ShallowWrapper {}`;
